Objects/unicodeobject.c

```diff
@@ -15083,86 +15260,162 @@ _PyUnicode_InternInPlace(PyInterpreterState *interp, PyObject **p)
+    {
+        PyObject *r = (PyObject *)_Py_hashtable_get(INTERNED_STRINGS, s);
+        if (r != NULL && r != s) {
+            assert(_Py_IsImmortal(r));
+            Py_SETREF(*p, Py_NewRef(r));
+            return;
+        }
+    }
 
-    /* Look in the per-interpreter cache. */
+    /* Do a setdefault on the per-interpreter cache. */
     PyObject *interned = get_interned_dict(interp);
     assert(interned != NULL);
 
     PyObject *t;
-    int res = PyDict_SetDefaultRef(interned, s, s, &t);
-    if (res < 0) {
-        PyErr_Clear();
-        return;
-    }
-    else if (res == 1) {
-        // value was already present (not inserted)
-        Py_SETREF(*p, t);
-        return;
+    {
+        int res = PyDict_SetDefaultRef(interned, s, s, &t);
+        if (res < 0) {
+            PyErr_Clear();
+            return;
+        }
+        else if (res == 1) {
+            // value was already present (not inserted)
+            s = t;
+            Py_SETREF(*p, t);
+            if (immortalize &&
+                    PyUnicode_CHECK_INTERNED(s) == SSTATE_INTERNED_MORTAL) {
+                immortalize_interned(s);
+            }
+            return;
+        }
+        else {
+            // value was newly inserted
+            assert (s == t);
+            Py_DECREF(t);
+        }
     }
-    Py_DECREF(t);
 
-    if (_Py_IsImmortal(s)) {
-        // XXX Restrict this to the main interpreter?
-        _PyUnicode_STATE(*p).interned = SSTATE_INTERNED_IMMORTAL_STATIC;
-        return;
-    }
+    /* NOT_INTERNED -> INTERNED_MORTAL */
 
+    assert(_PyUnicode_STATE(s).interned == SSTATE_NOT_INTERNED);
+
+    if (!_Py_IsImmortal(s)) {
+        /* The two references in interned dict (key and value) are not counted.
+        unicode_dealloc() and _PyUnicode_ClearInterned() take care of this. */
+        Py_SET_REFCNT(s, Py_REFCNT(s) - 2);
 #ifdef Py_REF_DEBUG
-    /* The reference count value excluding the 2 references from the
-       interned dictionary should be excluded from the RefTotal. The
-       decrements to these objects will not be registered so they
-       need to be accounted for in here. */
-    for (Py_ssize_t i = 0; i < Py_REFCNT(s) - 2; i++) {
+        /* let's be pedantic with the ref total */
         _Py_DecRefTotal(_PyThreadState_GET());
+        _Py_DecRefTotal(_PyThreadState_GET());
+#endif
+    }
+    _PyUnicode_STATE(s).interned = SSTATE_INTERNED_MORTAL;
+
+    /* INTERNED_MORTAL -> INTERNED_IMMORTAL (if needed) */
+
+#ifdef Py_DEBUG
+    if (_Py_IsImmortal(s)) {
+        assert(immortalize);
     }
 #endif
-    _Py_SetImmortal(s);
-    _PyUnicode_STATE(*p).interned = SSTATE_INTERNED_IMMORTAL;
+    if (immortalize) {
+        immortalize_interned(s);
+    }
+}
+
+void
+_PyUnicode_InternImmortal(PyInterpreterState *interp, PyObject **p)
+{
+    intern_common(interp, p, 1);
+}
+
+void
+_PyUnicode_InternMortal(PyInterpreterState *interp, PyObject **p)
+{
+    intern_common(interp, p, 0);
+}
```

Comment:
Likewise for these two functions.

```suggestion
PyObject *
_PyUnicode_InternImmortal(PyInterpreterState *interp, PyObject *s)
{
    return intern_common(interp, s, 1);
}

PyObject *
_PyUnicode_InternMortal(PyInterpreterState *interp, PyObject *s)
{
    return intern_common(interp, s, 0);
}
```

The same goes for `_PyUnicode_InternStatic()`.

Reply:
I was going to ask, "Why keep the in-place approach for these new *internal* C-API functions?"  However, on reflection, I'm not sure it matters that much either way.  Personally, I find the return-the-final-value approach to be more explicit, with less possibility for confusion at call sites and with less need to look up the behavior.  I'd lean toward taking that approach rather that "in place", but, again, it isn't that big a deal (especially for internal-only C-API).  Assuming I've explained myself well enough, I'll defer to your judgement.

Reply:
It's easier to replace `_PyUnicode_InternInPlace` with it, which helps backporting.